enhancements/ingress/client-tls.md

```diff
@@ -0,0 +1,325 @@
+// authentication, which can be used for mutual TLS.
+type ClientTLS struct {
+	// clientCertificatePolicy specifies whether the ingress controller
+	// requires clients to provide certificates.  This field accepts the
+	// values "Required" or "Optional".
+	//
+	// +kubebuilder:validation:Required
+	// +required
+	ClientCertificatePolicy ClientCertificatePolicy `json:"clientCertificatePolicy"`
+
+	// clientCA is an optional reference to a configmap containing the
+	// PEM-encoded CA certificate bundle that should be used to verify a
+	// client's certificate.  If omitted, the system trust bundle is used.
+	//
+	// +optional
+	ClientCA configv1.ConfigMapNameReference `json:"clientCA"`
+
+	// allowedSubjectPatterns specifies a list of regular expressions that
+	// should be matched against the distinguished name on a client
+	// certificate to filter requests.  If this list is empty, no filtering
+	// is performed.  If the list is nonempty, then at least one pattern
+	// must match a client certificate's distinguished name or else the
+	// ingress controller rejects the certificate and denies the connection.
+	//
+	// +optional
+	AllowedSubjectPatterns []string `json:"allowedSubjectPatterns,omitempty"`
+}
+```
+
+The `ClientCertificatePolicy` type accepts either one of two values: `Required`
+or `Optional`:
+
+```go
+// ClientCertificatePolicy describes the policy for client certificates.
+// +kubebuilder:validation:Enum="";Required;Optional
+type ClientCertificatePolicy string
+
+const (
+	// ClientCertificatePolicyRequired indicates that a client certificate
+	// should be required.
+	ClientCertificatePolicyRequired ClientCertificatePolicy = "Required"
+
+	// ClientCertificatePolicyOptional indicates that a client certificate
+	// should be requested but not required.
+	ClientCertificatePolicyOptional ClientCertificatePolicy = "Optional"
+)
+```
+
+If the user does not specify `spec.clientTLS`, then client TLS is not enabled,
+which means that the IngressController does not request client certificates on
+TLS connections.  If `spec.clientTLS` is specified, then the IngressController
+does request client certificates, and `spec.clientTLS.clientCertificatePolicy`
+must be specified to indicate whether the IngressController should reject
+clients that do not provide valid certificates.
+
+The optional `ClientCA` field may be used to specify a reference to a configmap
+that is in the same namespace as the IngressController and contains a CA
+certificate bundle.  If `ClientCA` is omitted, then the IngressController uses
+the system trust bundle.  If `ClientCA` is specified, then the IngressController
+uses the provided bundle to verify client certificates, and the
+IngressController does not start serving unless and until it can read the CA
+certificate bundle as well as the certificate revocation list (CRL) for each
+certificate therein that specifies one or more destination points for retrieving
+that certificate's CRL.
```

Comment:
On what basis?  This seems like a point of failure and it's not clear which way a user would prefer the system to behave.

This also doesn't appear to describe refreshes of this data.

Reply:
Presumably it is better not to start at all than to start without a CRL if one is configured.  If people care about client certificates and specify a CRL (either on the IngressController or on the CA certificate), I assume the would prefer the router not to start if the CRL cannot be retrieved.  

I can add a paragraph stating that the router will continue using the old CRL if the operator fails to retrieve updates.  Or maybe there should even be an option to stop serving traffic if the CRL remains unavailable for some period of time?

Reply:
Isn't this why Online Certificate Status Protocol (OCSP) is used in the industry? Why are we not building on that? 

In short if this verification happens at the route level; the OCSP responder could be responsible for updating the CRL or verification setting; in theory offloading the responsibility to an industry standard.